### PR TITLE
WIP: Rolldown build

### PR DESCRIPTION
**Pull request by underfin**
### E2E Test Status

windows ci has error
| Playground Name | Dev | Build | Reasons |
|--------|--------|--------|--------|
| alias | ✅ | ✅ |  |
| assets | ✅ | ✅  |  - disable rebuild test |
| assets-sanitize | ✅ |  ✅  | |
| backend-integration | ✅  | ❌ | https://github.com/rolldown/rolldown/issues/1842#issuecomment-2296345255 |
| build-old | ✅  | ✅  | - |
| cli | ✅ | ✅ | - |
| cli-module | ✅ | ✅  | - |
| config | ✅ | ✅ | - |
| csp | ✅ | ✅ | - |
| css | ✅ | ✅ | - |
| css-codesplit | ✅ | ❌ | manualChunks 
| css-codesplit-cjs | ✅ | ✅ | - 
| css-dynamic-import | ✅ | ❌ | https://github.com/rolldown/rolldown/issues/1842
| css-lightningcss | ✅ | ✅ | - |
| css-lightningcss-proxy | ✅ | ✅ | - |
| css-no-codesplit | ✅ | ✅  | - |
| css-sourcemap | ✅ | ✅ | - |
| data-uri | ✅ | ✅ | - |
| define | ✅ | ✅ | - |
| dynamic-import | ✅  | ❌ | https://github.com/rolldown/rolldown/issues/1842 https://github.com/rolldown/rolldown/issues/1843 |
| dynamic-import-inline | - | ✅ |  |
| env | ✅ | ✅ | - |
| env-nested | ✅ | ✅ | - |
| extensions | ✅ | ✅ | - |
| external | ✅ | ❌ | https://github.com/rolldown/rolldown/issues/2041 |
| fs-serve | ✅ | ✅ | - |
| glob-import | ✅ | ❌ | - remove empty CSS generate chunk https://github.com/rolldown/rolldown/issues/1842 |
| hmr | ✅ | ✅ | - |
| hmr-root | ✅  | - | - |
| hmr-ssr | ✅ | ✅ | - |
| html | ✅ | ✅ | - |
| import-assertion | ✅ | ✅ | - |
| js-sourcemap | ✅ | ❌ | manualChunks |
| json | ✅ | ✅ | - |
| legacy | ❌ | ❌ | 🟡pending support system format |
| lib | ❌ | ❌ | 🟡pending support umd format |
| minify | - |  ✅ | - |
| module-graph | ✅ | - | - |
| multiple-entrypoints | ✅ |  ✅ | - |
| nested-deps | ✅ | ✅ | - |
| object-hooks | ✅ | ❌ | object hook sequential |
| optimize-deps | ✅ |  ❌ | - the rolldown missing export always is error, but the rollup has error or warn level, ref https://github.com/rolldown/vite/pull/27/commits/bfc679efab9d3eb0ef3cee1611506518bcc33a23 <br> -https://github.com/rolldown/rolldown/issues/2031 <br> - using Array input generate name is not expected, the entry name will be like index~1.js |
| optimize-deps-no-discovery | ✅ | ✅ | - |
| optimize-missing-deps | ✅ | ✅ | - |
| preload | ✅ | ✅ | - |
| preserve-symlinks | ✅  | ✅  | - |
| proxy-bypass |  ✅ | ✅  | - |
| proxy-hmr | ✅ | ✅ | - |
| resolve | ✅  | ✅ |  |
| resolve-config | - | ✅   |  |
| ssr | ✅ | ✅ | - |
| ssr-alias | ✅ | ✅ | - |
| ssr-conditions | ✅ | ✅ | - |
| ssr-deps | ✅ | ✅ | - |
| ssr-html | ✅ | ✅ | - |
| ssr-noexternal | ✅ | - | - |
| ssr-pug | ✅ | ✅ | - |
| ssr-resolve | - | ✅ | - |
| ssr-webworker | ✅ | ✅ | |
| tailwind | ✅ | ✅ | - |
| tailwind-sourcemap | ✅  | ✅  | - |
| transform-plugin | ✅ | ✅ | |
| tsconfig-json | ✅ | ✅ | - |
| tsconfig-json-load-error | ✅ | ✅ | - |
| wasm | ✅ | ✅ | - |
| worker | ✅ | ❌ | -https://github.com/rolldown/rolldown/issues/1842 <br> -https://github.com/rolldown/rolldown/issues/1010 |

<details>
<summary>
hook call cost for benchcases/20000modules
</summary>

```bash
vite:esbuild
  |- transform: 25867596.34ms
  |- load: 1.68ms
  |- transform_ast: 0.00ms
  |- resolve: 0.00ms
vite:load-fallback
  |- load: 7975644.26ms
  |- transform: 7.36ms
  |- transform_ast: 0.02ms
  |- resolve: 0.00ms
alias
  |- resolve: 6765733.93ms
  |- load: 0.52ms
  |- transform: 0.05ms
  |- transform_ast: 0.00ms
vite:modulepreload-polyfill
  |- load: 2099766.14ms
  |- resolve: 196862.31ms
  |- transform: 0.16ms
  |- transform_ast: 0.00ms
vite:css
  |- transform: 1300225.05ms
  |- load: 67072.74ms
  |- transform_ast: 0.00ms
  |- resolve: 0.00ms
vite:resolve
  |- resolve: 644735.49ms
  |- load: 564377.40ms
  |- transform: 0.10ms
  |- transform_ast: 0.00ms
vite:json
  |- transform: 455512.22ms
  |- load: 0.70ms
  |- transform_ast: 0.02ms
  |- resolve: 0.00ms
vite:wasm-helper
  |- load: 338319.71ms
  |- transform: 2.35ms
  |- transform_ast: 0.01ms
  |- resolve: 0.00ms
vite:dynamic-import-vars
  |- load: 262145.11ms
  |- transform: 60529.64ms
  |- transform_ast: 0.01ms
  |- resolve: 0.00ms
vite:build-import-analysis
  |- load: 93937.80ms
  |- transform: 92939.95ms
  |- transform_ast: 0.01ms
  |- resolve: 0.00ms
vite:worker
  |- transform: 82788.87ms
  |- load: 68149.93ms
  |- transform_ast: 0.01ms
  |- resolve: 0.00ms
vite:asset
  |- load: 99090.04ms
  |- transform: 1.84ms
  |- transform_ast: 0.02ms
  |- resolve: 0.00ms
vite:define
  |- transform: 89704.89ms
  |- load: 1.10ms
  |- transform_ast: 0.00ms
  |- resolve: 0.00ms
vite:wasm-fallback
  |- load: 65423.90ms
  |- transform: 0.81ms
  |- transform_ast: 0.02ms
  |- resolve: 0.00ms
vite:html-inline-proxy
  |- load: 62983.32ms
  |- transform: 0.03ms
  |- transform_ast: 0.01ms
  |- resolve: 0.00ms
vite:css-post
  |- transform: 50803.16ms
  |- load: 0.48ms
  |- transform_ast: 0.01ms
  |- resolve: 0.00ms
vite:reporter
  |- transform: 48866.64ms
  |- load: 1.36ms
  |- resolve: 0.00ms
  |- transform_ast: 0.00ms
vite:asset-import-meta-url
  |- transform: 42374.47ms
  |- load: 0.31ms
  |- transform_ast: 0.02ms
  |- resolve: 0.00ms
vite:build-html
  |- transform: 41530.89ms
  |- load: 0.40ms
  |- transform_ast: 0.01ms
  |- resolve: 0.00ms
vite:worker-import-meta-url
  |- transform: 35687.16ms
  |- load: 0.49ms
  |- transform_ast: 0.00ms
  |- resolve: 0.00ms
vite:build-metadata
  |- load: 4.54ms
  |- transform: 2.74ms
  |- resolve: 1.78ms
  |- transform_ast: 0.01ms
glob_import_plugin
  |- transform_ast: 1.79ms
  |- transform: 0.44ms
  |- load: 0.40ms
  |- resolve: 0.00ms
rolldown:data-url
  |- transform: 1.25ms
  |- transform_ast: 0.02ms
  |- resolve: 0.00ms
  |- load: 0.00ms
vite:watch-package-data
  |- load: 0.32ms
  |- resolve: 0.14ms
  |- transform: 0.08ms
  |- transform_ast: 0.00ms

```
</details>

### CODE REVIEW

**Comment by sapphi-red** [comment]
After bumping rolldown to 0.13.2 locally, I encountered this bug: https://github.com/rolldown/rolldown/issues/2183
It's strange that the reproduction doesn't work even though the test in this branch passes with the version before bumping (`0.12.2-snapshot-aa5807f-20240902003235`).

**Comment by underfin** [comment]
> After bumping rolldown to 0.13.2 locally, I encountered this bug: [rolldown/rolldown#2183](https://github.com/rolldown/rolldown/issues/2183) It's strange that the reproduction doesn't work even though the test in this branch passes with the version before bumping (`0.12.2-snapshot-aa5807f-20240902003235`).

What case you meet it, I can't found it.

**Comment by sapphi-red** [comment]
> > After bumping rolldown to 0.13.2 locally, I encountered this bug: [rolldown/rolldown#2183](https://github.com/rolldown/rolldown/issues/2183) It's strange that the reproduction doesn't work even though the test in this branch passes with the version before bumping (`0.12.2-snapshot-aa5807f-20240902003235`).
> 
> What case you meet it, I can't found it.

This test case failed because `sub_1.css` didn't exist.
https://github.com/rolldown/vite/blob/75371b9173d11116a6f775313c88325458ee7e51/playground/css/__tests__/same-file-name/css-same-file-name.spec.ts#L15-L17

This is the branch I'm testing on: https://github.com/rolldown/vite/compare/rolldown-build...rolldown:vite:rolldown-build-investigate (note that `./playground/css/__tests__/same-file-name/*.spec.ts` is excluded in that branch)

**Comment by underfin** [comment]
Yeah. It's strange and I also can't found it at now, the test how to work 😭. I remember i changed it because here has a filename different and all file is exits...

File: packages/vite/src/node/plugins/asset.ts [hunk 1]
```diff
@@ -31,9 +31,10 @@ import {
 import { DEFAULT_ASSETS_INLINE_LIMIT, FS_PREFIX } from '../constants'
 import type { ModuleGraph } from '../server/moduleGraph'
 import { cleanUrl, withTrailingSlash } from '../../shared/utils'
+import { getChunkMetadata } from './metadata'
 
-// referenceId is base64url but replaces - with $
-export const assetUrlRE = /__VITE_ASSET__([\w$]+)__(?:\$_(.*?)__)?/g
+// Not sure why rollup referenceId is base64url but replaces - with $ at https://github.com/rollup/rollup/pull/5155
+export const assetUrlRE = /__VITE_ASSET__([\w-]+)__(?:\$_(.*?)__)?/g
```
[the remark this answers is not in the document]
Comment (underfin) [reply packages/vite/src/node/plugins/asset.ts:37]: Ok. Thank you get it.

File: packages/vite/src/node/plugins/terser.ts [hunk 2]
```diff
@@ -94,8 +94,8 @@ export function terserPlugin(config: ResolvedConfig): Plugin {
       }
     },
 
-    closeBundle() {
-      worker?.stop()
-    },
+    // closeBundle() {
+    //   worker?.stop()
+    // },
```
[the remark this answers is not in the document]
Comment (underfin) [reply packages/vite/src/node/plugins/terser.ts:99]: Yeah. We could implement it using an empty function. I will discuss it at meeting.

[the remark this answers is not in the document]
Comment (underfin) [reply packages/vite/src/node/plugins/terser.ts:99]: Implement it at https://github.com/rolldown/rolldown/pull/2197

**Comment by sapphi-red** [comment]
Fixing https://github.com/rolldown/rolldown/issues/2183 unveiled a different bug: https://github.com/rolldown/rolldown/issues/2207

**Comment by underfin** [comment]
> Fixing [rolldown/rolldown#2183](https://github.com/rolldown/rolldown/issues/2183) unveiled a different bug: [rolldown/rolldown#2207](https://github.com/rolldown/rolldown/issues/2207)

Thank you. I will fixed it at next.

**Comment by sapphi-red** [comment]
Closing as I rebased this PR on the main branch in #43

**Pull request closed without merging**